src/Compilers/Core/Portable/Serialization/ObjectWriter.cs

```diff
@@ -482,6 +496,19 @@ private void WriteArray(Array instance)
 
         private void WriteType(Type type)
         {
+            // optimization. primitive types
+            if (type == typeof(byte))
```

Comment:
consider breaking out the reader/writer work into its own PR so you can get it in separately and make this PR shorter.